### ISSUE TITLE
Check missing packages beforehand

### ISSUE THREAD

**Issue by chrarnold** [mlr-org/mlr3#432]
Hi guys,thanks for mlr3, it is super nice. I started using it and I encountered some things that I think can be improved.
For benchmark_grid, when giving a list of learners, it is not tested before starting whether the packages for the learners are actually installed. 
1) This may result in error messages after some learners finished, causing wasted computation time because one has to restart. It should be checked before doing anything.
2) If some learners finished while for others, the package could not be loaded, can the intermediate results be stored so they don't have to be recomputed?

**Comment by mllg** [comment]
> For benchmark_grid, when giving a list of learners, it is not tested before starting whether the packages for the learners are actually installed.

We can also `requireNamespace()` the learner package on construction. OTOH, if we consider parallel computation, we do not necessarily need the learner packages installed on the master, only on the slaves. Maybe print a warning if the package is not installed?

> If some learners finished while for others, the package could not be loaded, can the intermediate results be stored so they don't have to be recomputed?

If you encapsulate the learners, e.g. with 
```r
learners$encapsulate = c(train = "evaluate")
```
, you should already be storing partial results. You can also quite easily find out which learners failed with:
```r
mlr3misc::map_lgl(bmr$data$learner, function(x) length(x$errors) > 0)
```

So the missing part is to trigger a re-computation of the respective experiments. This is definitely doable, but not high prio at the moment. I'll open a separate issue for this.

**Comment by berndbischl** [comment]
> We can also `requireNamespace()` the learner package on construction. OTOH, if we consider parallel computation, we do not necessarily need the learner packages installed on the master, only on the slaves. Maybe print a warning if the package is not installed?

i get what you mean, but isnt supporting the case "have packakes not on the master but only slave" a quite exotic edge case? while the user here runs into a pretty normal case, then loses his results? i would rather guard against this problem now, seems quite realistic to me

**Comment by berndbischl** [comment]
> If you encapsulate the learners, e.g. with
> 
> ```r
> learners$encapsulate = c(train = "evaluate")
> ```
> 
> , you should already be storing partial results. You can also quite easily find out which learners failed with:
> 
> ```r
> mlr3misc::map_lgl(bmr$data$learner, function(x) length(x$errors) > 0)
> ```
> 
> So the missing part is to trigger a re-computation of the respective experiments. This is definitely doable, but not high prio at the moment. I'll open a separate issue for this.

I agree that the we should dicuss how to do this first. 

f you are saying that partial result access is already available, where is this documented? Shouldnt we then not demonstrate this in the errorhandling section of the book?

**Comment by mllg** [comment]
> i get what you mean, but isnt supporting the case "have packakes not on the master but only on slave" during parallelization a quite exotic edge case? while the user here runs into a pretty normal case, then loses his results? i would rather guard against this problem now, seems quite realistic to me

Yes it is an exotic edge case, and yes, we should probably signal something before starting the benchmark. I would suggest signaling a warning on learner construction. This way, you can tryCatch/suppress/muffle the warning if you need to, but still can construct the learner.



> f you are saying that partial result access is already available, where is this documented? Shouldnt we then not demonstrate this in the errorhandling section of the book?

Well, this is of limited use if you cannot "repair" the benchmark by restarting subsets of experiments. Learners may fail, we have encapsulation and fallback learners for this - and this is documented.

**Comment by mllg** [comment]
You will now get a warning if you construct a task generator, a learner or a measure w/o the required packages installed.

**Comment by berndbischl** [comment]
> You will now get a warning if you construct a task generator, a learner or a measure w/o the required packages installed.

great. that at least gives users a very fair channce

the other issue of generating "partial" results is more complex and  we should discuss

**Comment by berndbischl** [comment]
> Well, this is of limited use if you cannot "repair" the benchmark by restarting subsets of experiments. Learners may fail, we have encapsulation and fallback learners for this - and this is documented.

no hang own. why is this of limited use? the OP was asking about this, can you at least point to the documention that explains the generation of partial results? also the BMR will still be valid 
?

**Comment by chrarnold** [comment]
Thanks guys, and thanks for the quick fix. I am always trying to think of a (less experienced) user perspective, and I think mlr3 can still improve a bit :). Regarding partial results: If this already works, why not add it to the mlr3 documentation / book, somewhere at least. I have a few more practical experiences from going from an (advanced R) mlr3 novice to a "I run a mlr3 analyses successfully" user that I could share if necessary. You might remember me from the EMBL course last year, maybe a phone call might be better to provide the feedback otherwise of course here.

**Comment by berndbischl** [comment]
> I am always trying to think of a (less experienced) user perspective, and I think mlr3 can still improve a bit :

you posted a very valid concern. pls post more.



> Regarding partial results: If this already works, why not add it to the mlr3 documentation / book, somewhere at least

agree. see my post above

> I have a few more practical experiences from going from an (advanced R) mlr3 novice to a "I run a mlr3 analyses successfully" user that I could share if necessary.

please really do this, this is more then welcome. if you need better communication, you can also go to our mattermost

**Comment by mllg** [comment]
To be sure we are all on the same page, this is what I mean by "partial results" (here for resampling, you can do the same for benchmarks):

```r
library(mlr3)
task = tsk("sonar")

# learner which fails randomly with probability 30%
learner = lrn("classif.debug", error_train = 0.3)

# encapsulate training in evaluate-wrapper
learner$encapsulate = c(train = "evaluate")

set.seed(123)
rr = resample(task, learner, rsmp("cv"))

# table of logged errors
rr$errors

# available scores
rr$score(msr("classif.acc"))$classif.acc

# partial RR data
as.data.table(rr)[!rr$errors, on = "iteration"]
```

From a statistical point of view, it is unclear how to proceed from here. Aggregation over the subset of successfully calculated performance values is often not sound (you would have to make the assumption that the calculation failed completely at random, and not because the splits into train/test were more or less difficult than others).

If you cannot just omit the missing values, you still can impute a performance value, but the value is often unclear. E.g., for the accuracy you could opt for a value of 0, but this penalization often seems to harsh. For other measures, e.g. MSE, it is completely unclear what to do.

This is why we implemented fallback learners which are fitted if the original learner failed to train or predict. If you now fallback to a featureless learner, you predict the majority class (classification) or mean (regression) based on the same train/test split the original learner failed to fit on. This way, you get a good baseline (even for unbalanced classification problems), and the aggregation over the performances makes sense again. You can of course also fallback to something more sophisticated, if this makes sense in your setup.

So the recalculation of failed resampling iterations (or parts of benchmarks) probably only makes sense if you assume that the learner failed due to temporary problems with the CPU, network, installation, etc. I've opened #433 for this.

**Comment by berndbischl** [comment]
> From a statistical point of view, it is unclear how to proceed from here. [snip]
> This is why we implemented fallback learners [snip]

completely agree. (not surprising as we discussed this at length and designed the above principle as the best we could come up with for (statistical) benchmarking)
i do hope we describe what you wrote above like this in the book - because its a good explanation

> So the recalculation of failed resampling iterations (or parts of benchmarks) probably only makes sense

BUT: 
here and especially in the first paragraph you are not answering the what the user requests - at least in my reading and thinking.

i think what the user wants (or maybe i myself here?) is some form of "checkpointing" or permanent storage. there can be many reasons for this. maybe you simply want to peak into what you conputed to understand the situation better. maybe to impute missing results later manually. there can be many questions. he is asking about a particular technical feature - you are saying: "you have to approach it in this complete process, and then it might not even be necessary what you wanted in the first place".

isnt it so unrealistic to ask about filing-to-disk of results in such situations? think about batchtools, does it too (yes also for communication purposes)
whether we want to implement something like this immediatly, or in what form is another question. but at least i can see some appeal.

@chrarnold i hope i did not misrepresent what you are asking. otherwise maybe explain a concrete  option / proposal / way what you are ideally looking for

**Comment by chrarnold** [comment]
Hi Bernd,
just to clarify what I originally meant: I had a case this week where I was running multiple learners in a benchmarking framework. A few of them finished after a while, but then I got the "package not found" error message, at which point the calculation stopped at the final results of the learner was lost. That was my original point of storing intermediate results, sorry if this was unclear.  With the new "warnings" fix, this will still happen, but at least there was a warning. The warning could be explicit like "You may lose results from learners that are finished successfully before because of the missing package" so the user would not even (if he or she is in a clear state of mind :) ) start the training process at all - with the warning as of now, it might be like "yeah so", not thinking about that hours of computation might be lost because of the missing package. Only if the missing package is in fact the first learner, there wouldn't be consequences, as the function stops right away.

**Comment by berndbischl** [comment]
> Hi Bernd,
> just to clarify what I originally meant: I had a case this week where I was running multiple learners in a benchmarking framework. A few of them finished after a while, but then I got the "package not found" error message, at which point the calculation stopped and the final results of the learner was lost. That was my original point of storing intermediate results (intermediate in the sense of finished results from learners in the benchmark setting), sorry if this was unclear.

that was super clear to me for the whole thread 

> With the new "warnings" fix, this will still happen, but at least there was a warning

yes

i think we are discussing 2 different issues here. 

- the guarding / warning. thats now solved.
- the storing of disk - that i posted about above

would you like such permanent results on disk? even after we have implemented the warning? or dont you see the use and i am misunderstanding you?

**Comment by chrarnold** [comment]
All I am saying was that for the case I had - a few learners finished successfully, and the 4th or 5th one failed with a package related error, I had to redo the first ones completely, which is a bit unfortunate. I understand your concern though - how in practise this could be stored. My case might be a rather special one, which is why I feel a clear and explicit warning as suggested above might be sufficient, so nothing else needs to be done, no complicated storing of intermediate results if not super straight-forward. I think a super-clear warning is probably enough for now - I imagine that the training process usually doesn't end up in errors, then storing intermediate results might not be needed for now.

**Comment by berndbischl** [comment]
> . I understand your concern though

i have currently no concerns at all. just trying to understand the situation. TBH i dont think it would be that hard to store benchmark results on disk. and maybe that would even be good.

> I imagine that the training process usually doesn't end up in errors,

it unfortunately does. the larger the benchmark, the higher the risk. this is why we implemented the evaluate-mode, the fallback-learner, and why i am asking here. it is quite an important point

**Comment by pat-s** [comment]
> All I am saying was that for the case I had - a few learners finished successfully, and the 4th or 5th one failed with a package related error, I had to redo the first ones completely, which is a bit unfortunate. 

That's why I use {drake} for all my analyses with {mlr3}. It does exactly this and also aggregates the models in the end for you again. This requires running one `benchmark()` call for each learner (which is then called a "target" in {drake}) and aggregate the results in the end. See [mlr3-learndrake](https://github.com/mlr-org/mlr3-learndrake) for a template repo.

Regarding supporting this natively in {mlr3}: I am really skeptical if we want to go down this route ourselves. There are already existing tools that do this very well and its more complicated than it may seem at first.

---

Regarding the new warning about missing packages:

Why not add a `menu()` call asking the user if one really wants to continue even though we detected missing packages which might cause learners to fail?
